### PR TITLE
Add split() method to BufferedUarte in embassy-nrf

### PR DESCRIPTION
**Pull request by ZoeyR**
I haven't completed testing this yet. I'm creating this PR early so that I can get corrected if I went way off course.

This PR adds a `split()` method to `BufferedUarte` as discussed on matrix.

### CODE REVIEW

**Comment by Dirbaio** [comment]
LGTM overall.

It'd be nice if the write/read code wasn't duplicated. Perhaps you can make the Tx/Rx structs have a `&BufferedUarte`, and add `read`, `write` async fns to `BufferedUarte` that take `&self`, and then call these from all the trait impls?

**Comment by ZoeyR** [comment]
Yeah, that should be workable. I assume for now we'd want those to be internal to the module?

**Comment by Dirbaio** [comment]
yes

**Comment by ZoeyR** [comment]
Factored those out!

File: embassy-nrf/src/buffered_uarte.rs [hunk 1]
```diff
@@ -232,22 +227,64 @@ impl<'d, U: UarteInstance, T: TimerInstance> embedded_io::asynch::Read for Buffe
...
     }
-}
 
-impl<'d, U: UarteInstance, T: TimerInstance> embedded_io::asynch::BufRead for BufferedUarte<'d, U, T> {
-    type FillBufFuture<'a> = impl Future<Output = Result<&'a [u8], Self::Error>>
-    where
-        Self: 'a;
+    async fn inner_write<'a>(&'a self, buf: &'a [u8]) -> Result<usize, core::convert::Infallible> {
+        poll_fn(move |cx| {
+            let res = self.inner.borrow_mut().with(|state| {
```
Comment (Dirbaio) [inline embassy-nrf/src/buffered_uarte.rs:240]: avoid borrowing twice: `let inner = self.inner.borrow_mut();` then use `inner`.

Comment (ZoeyR) [reply embassy-nrf/src/buffered_uarte.rs]: I did it for the functions that double-borrowed. I left the ones with a single borrow intact. Should I switch those too?

**Review by Dirbaio — APPROVED** [review]
LGTM, feel free to merge yourself when you've tested it out! Thanks :)

bors d+

**Comment by ZoeyR** [comment]
Okay, did some testing and it seems to work!

bors r+

**Pull request merged**